File: 백준/Gold/1766. 문제집/문제집.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int[] inDegree = new int[n+1];
        Map<Integer, List<Integer>> adj = new HashMap<>();

        for(int i = 0; i < m; i++) {

            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            adj.putIfAbsent(a, new ArrayList<>());
            adj.get(a).add(b);
            inDegree[b]++;
        }

        PriorityQueue<Integer> pq = new PriorityQueue<>();
        List<Integer> result = new ArrayList<>();

        for(int i = 1; i <= n; i++) {
            if(inDegree[i] == 0) pq.add(i);
        }

        while(!pq.isEmpty()) {
            int cur = pq.poll();
            result.add(cur);

            if(adj.containsKey(cur)) {
                for(int nxt : adj.get(cur)) {
                    inDegree[nxt]--;
                    if(inDegree[nxt] == 0) pq.add(nxt);
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        result.forEach(num -> sb.append(num + " "));
        System.out.println(sb);
    }
}

File: 백준/Gold/1766. 문제집/문제집.kt
import java.io.*
import java.util.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

        fun String.parseInts() = split(" ").map{ it.toInt() }
        val (n ,m) = readLine().parseInts()

        val inDegree = IntArray(n + 1) { 0 }
        val adj = mutableMapOf<Int, MutableList<Int>>()

        repeat(m) {
            val (a, b) = readLine().parseInts()
            adj.computeIfAbsent(a) { mutableListOf() }.add(b)
            inDegree[b]++
        }

        fun topologicalSort(): List<Int> {

            val pq = PriorityQueue<Int>()
            val result = mutableListOf<Int>()

            (1..n)
                .filter { inDegree[it] == 0 }
                .forEach { pq.offer(it) }

            generateSequence { pq.poll() }
                .takeWhile { it != null }
                .forEach { cur ->
                    result.add(cur)
                    adj[cur]?.forEach { next ->
                        inDegree[next]--
                        if(inDegree[next] == 0) pq.offer(next)
                    }
                }

            return result
        }

        topologicalSort()
            .joinToString(" ")
            .let { println(it) }
    }